### DISCUSSION TITLE
Github Actions interprets non-zero Python exit code as success

### DISCUSSION THREAD

**Discussion by kasuteru** [community/community#57934] [Actions] [answered]
### Select Topic Area

Bug

### Body

According to the documentation, a non-zero exit code is considered a failure of a github action step: [link](https://docs.github.com/en/actions/creating-actions/setting-exit-codes-for-actions#about-exit-codes)

However, my application returns exit code 9009 in case of failure (this number was chosen by the library I am using, not by myself). Github Actions reports this as success even though it is clearly non-zero.

Of course, I can work around this by try...excepting my whole script and forcing it to return 1 - but this would be just a manual workaround. Certainly, this is a bug / unexpected behavior?

**Comment by airtower-luna** [comment, marked as answer]
That is pretty clearly a bug in the library you're using. The [`sys.exit()` documentation](https://docs.python.org/3/library/sys.html#sys.exit) warns about things like this (emphasis added):

> If it is an integer, zero is considered “successful termination” and any nonzero value is considered “abnormal termination” by shells and the like. **Most systems require it to be in the range 0–127, and produce undefined results otherwise.**

On Linux on x86_64 the *actual* return code I receive after `sys.exit(9009)` is 49: 9009 modulo 256, because [the least significant byte of the `exit()` value is used](https://manpages.debian.org/bullseye/manpages-dev/exit.3.en.html). But you shouldn't rely on that in Python code either, "undefined results" and all that. :sweat_smile: 

It's also possible that something in your workflow eats the exit code, but that's impossible to say for sure without seeing the workflow.

**Reply by kasuteru** [reply]
Thanks for that detailed answer! I was running this on a Windows machine, didn't realize that Linux actually takes the modulo.

On further investigation, this status code is not actually coming from Python but rather being translated by a Python library from a networking client error when calling an external storage. So I guess I have to use the workaround mentioned, since I have no way of controlling what number is sent by the external storage API.
